independent-projects/bootstrap/runner/src/main/java/io/quarkus/bootstrap/runner/JarFileReference.java

```diff
@@ -0,0 +1,178 @@
+            // The new file reference has been successfully published and can be used by the current and other threads
+            return consumeSharedJarFile(newJarFileRef, jarResource, resource, fileConsumer);
+        }
+
+        // The newly created file reference hasn't been published, so it can be used exclusively by the current virtual thread
+        return consumeUnsharedJarFile(newJarFileRef, jarResource, resource, fileConsumer);
+    }
+
+    private static <T> T consumeSharedJarFile(CompletableFuture<JarFileReference> jarFileReferenceFuture,
+            JarResource jarResource, String resource, JarFileConsumer<T> fileConsumer) {
+        JarFileReference jarFileReference = jarFileReferenceFuture.join();
+        try {
+            return fileConsumer.apply(jarFileReference.jarFile, jarResource.jarPath, resource);
+        } finally {
+            jarFileReference.release(jarResource);
+        }
+    }
+
+    private static <T> T consumeUnsharedJarFile(CompletableFuture<JarFileReference> jarFileReferenceFuture,
+            JarResource jarResource, String resource, JarFileConsumer<T> fileConsumer) {
+        JarFileReference jarFileReference = jarFileReferenceFuture.join();
+        try {
+            return fileConsumer.apply(jarFileReference.jarFile, jarResource.jarPath, resource);
+        } finally {
+            boolean closed = jarFileReference.release(jarResource);
+            assert !closed;
+            // Check one last time if the file reference can be published and reused by other threads, otherwise close it
+            if (!jarResource.jarFileReference.compareAndSet(null, jarFileReferenceFuture)) {
+                closed = jarFileReference.release(jarResource);
+                assert closed;
+            }
+        }
+    }
+
+    private static CompletableFuture<JarFileReference> syncLoadAcquiredJarFile(JarResource jarResource) {
+        try {
+            return CompletableFuture.completedFuture(new JarFileReference(JarFiles.create(jarResource.jarPath.toFile()), true));
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to open " + jarResource.jarPath, e);
+        }
+    }
+
+    private static CompletableFuture<JarFileReference> asyncLoadAcquiredJarFile(JarResource jarResource) {
+        CompletableFuture<JarFileReference> newJarFileRef;
+        do {
+            final var jarFileRefFuture = getJarFileReferenceFuture(jarResource);
+            newJarFileRef = jarFileRefFuture == null ? jarResource.jarFileReference.get() : jarFileRefFuture;
+        } while (!newJarFileRef.join().acquire());
+        return newJarFileRef;
+    }
+
+    private static CompletableFuture<JarFileReference> getJarFileReferenceFuture(JarResource jarResource) {
+        CompletableFuture future = new CompletableFuture();
+        if (jarResource.jarFileReference.compareAndSet(null, future)) {
+            future.completeAsync(() -> {
```

Comment:
why not "complete" here?

Reply:
![image](https://github.com/quarkusio/quarkus/assets/13125299/0fc70122-cc6b-4935-a1eb-5c9329ac1683)

this is going to use the "common FJ pool" I don't think is what we want here